### ISSUE TITLE
Enable CORS in ui.interaction endpoint

### ISSUE THREAD

**Issue by albertosgz** [RocketChat/Rocket.Chat#21033]
### Description:

Hi @sampaiodiego ! I hope you can help me with this.

I have an issue implementing the UI Kit library.

I have an error calling to `api/apps/ui.interaction/c33fa1a6-68a7-491e-bf49-9d7b99671c48` endpoint: 
```
Access to XMLHttpRequest at 'http://chat.local:3300/api/apps/ui.interaction/c33fa1a6-68a7-491e-bf49-9d7b99671c48' from origin 'http://localhost:8000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource
```

I see I can add cors easily into the code: 
https://github.com/RocketChat/Rocket.Chat/blob/e802fa59ce48fdee93a5ef56136b076acb213d87/app/apps/server/communication/uikit.js#L13
Whit this code:
https://expressjs.com/en/resources/middleware/cors.html

Also I see in the configuration of the RC Server that I can enable or disable the CORS in REST API, but that setting does not apply in this `ui.interaction` endpoint, right?

I would like to enable the CORS as same as REST API configuration.
I can make a PR, but I don't know how to get such configuration from `uikit.js` file. Can you point me in the right direction?

Thanks a lot!

### Steps to reproduce:

1. Enable CORS in REST Configuration.
2. I can query to `http://chat.local:3300/api/v1/method.call/slashCommand` but not to  `http://chat.local:3300/api/apps/ui.interaction/c33fa1a6-68a7-491e-bf49-9d7b99671c48`

### Expected behavior:

To works with CORS enable as same as REST API `api/v1/...`

### Actual behavior:

![image](https://user-images.githubusercontent.com/12124175/110257330-b9479000-7f9d-11eb-9dde-00b1a74ec83d.png)

### Server Setup Information:

Server Setup Information
Versión
3.11.1
ID de despliegue
vcohFpCn68jhkCkHT
Versión de Motor de Aplicaciones
1.22.2
Versión de Node
v12.18.4
Migración de base de datos
215 (7 de marzo de 2021 10:47
MongoDB
4.0.23 / mmapv1 (oplog Habilitado)
Detalles de la transacción
HEAD: (7c5d6ae2a)
Rama: HEAD
PID
10

### Version of Rocket.Chat Server

3.11.1

### Client Setup Information

- Desktop App or Browser Version: Chrome. Custom client app.
- Operating System: LINUX

### Additional context

I am developing a custom client to integrate with RC Server. And we want to connect with Poll app. So I need to implement the UI Kit.

**Comment by lolimay** [comment]
Hello, @albertosgz ! We provide this [slack compatible package](https://github.com/RocketChat/slack-compatibility-for-apps) for you guys to develop apps the slack way (a.k.a. access UIKit from outside Rocket.Chat)

**Comment by albertosgz** [comment]
Thanks @lolimay . I am not sure if SCL is what I need, since is not migrating an app from Slack, nor having a channel among them (slack <-> RocketChat).

However, with that documentation I see I can declare custom endpoints. But I see 2 downsides:
- I need to implement the whole RC App, when I would like to reuse the [Poll command](https://docs.rocket.chat/guides/rocket-chat-apps/poll), which I think is pretty cool, and doing so would be reinventing the wheel.
- It is pending to cover the issue that my app is not running in same domain as rocket chat server, so the Cross-Domain request is still happening, and then a CORS error.

**Comment by lolimay** [comment]
> However, with that documentation I see I can declare custom endpoints. But I see 2 downsides:
> 
> * I need to implement the whole RC App, when I would like to reuse the [Poll command](https://docs.rocket.chat/guides/rocket-chat-apps/poll), which I think is pretty cool, and doing so would be reinventing the wheel.
> * It is pending to cover the issue that my app is not running in same domain as rocket chat server, so the Cross-Domain request is still happening, and then a CORS error.

Regarding the point 1, yes, you still need write a little code to create a Rocket.Chat App to trigger UIKit actions since we don't allow apps/apis to trigger a UIKit (opening a modal for example) without users ' operations. So a UIKit action must be triggered by a user interaction (either by entering a slashcommand or by clicking a button etc.).

As for the point 2, CORS only works in the browser. A Rocket.Chat app is running on the server side and it can register some HTTP endpoints. Your app which is running on another server can communite with these endpoints to use UIKit indirectly. They are both working on the server side which will not get blocked by the CORS policy.

Hope these can make senses 🙈

**Comment by albertosgz** [comment]
Thanks for your reply @lolimay .

About point 1, yes, in my app the UIKit is trigger by the user with the request to `https://chat-test.ichair.org/api/v1/method.call/slashCommand`. Notice that for this endpoint, `.../api/v1/...`, the REST Configuration of this picture applies:
![image](https://user-images.githubusercontent.com/12124175/110347968-720adf00-8031-11eb-86d7-4254f4fd3cc0.png). 
Instead, the api I am finding issues is another one: `.../api/apps/ui.interaction/...`, which I guess has another configuration (is what I found in the code, in my first comment).

About point 2, I am in a browser, since my app is a chat in VueJS (based on this beautiful [code](https://github.com/antoine92190/vue-advanced-chat)). So CORS applies to me.

I was able to fix it with a workaround, setting up the nginx (acting as reverse proxy), adding next lines:
```
more_set_headers 'Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, X-User-Id, X-Auth-Token, x-visitor-token, Authorization';
more_set_headers 'Access-Control-Allow-Methods: GET, POST, PUT, DELETE, HEAD, PATCH';
more_set_headers 'Access-Control-Allow-Origin: *';
if ($request_method = 'OPTIONS') {
        # Tell client that this pre-flight info is valid for 20 days
        add_header 'Access-Control-Max-Age' 1728000;
        add_header 'Content-Type' 'text/plain charset=UTF-8';
        add_header 'Content-Length' 0;
        return 204;
}
```
Based on this [gist](https://gist.github.com/Stanback/7145487).

But it is still not configurable like for the endpoints `.../api/v1/...`. That's why I would like to add cors configuration to the endpoint `.../api/apps/ui.interaction/...` as I suggest in the first comment of the thread.

**Comment by lolimay** [comment]
cc @d-gubert

**Comment by albertosgz** [comment]
hi guys! any new on this? @lolimay

**Comment by jonathantyar** [comment]
any update on this? can we push it to the new build?

**Comment by lolimay** [comment]
> any update on this? can we push it push to new build?

It has been added to 3.14.0 and will be published soon.

**Comment by albertosgz** [comment]
@lolimay was included in the 3.14 release?

**Comment by lolimay** [comment]
Hi @albertosgz ! I am sorry it's hasn't been delivered into 3.14.0 release due to extra code reviews. I'll remind guys who is working this.

**Comment by jonathantyar** [comment]
I had workaround for that, needed to add custom configuration Caddyfile. if you guys interested i can share it, but please consider security reason

**Comment by jonathantyar** [comment]
```
{{yourrocketchaturl}} {

	@option {
	  method OPTIONS
	}

	route {
	  handle @option {
		header {
		  Access-Control-Allow-Origin *
		  Access-Control-Allow-Headers "Access-Control-Allow-Origin, Access-Control-Allow-Headers, x-requested-with, Content-Type, origin, authorization, accept, client-security-token, user-agent, x-auth-token, x-user-id"
		}
	    respond 200 {
	      close
	    }
	  }
	}

	reverse_proxy /* localhost:3000 {
	  header_down Access-Control-Allow-Origin *
	  header_down Access-Control-Allow-Methods "POST, GET, DELETE, PUT, HEAD, PATCH"
	  header_down Access-Control-Allow-Headers "Access-Control-Allow-Origin, Access-Control-Allow-Headers, x-requested-with, Content-Type, origin, authorization, accept, client-security-token"
	}
}
```
Caddyfile version 2.3.0

Remember, different version of Caddy will not work!

**Comment by sampaiodiego** [comment]
closed by #22150
